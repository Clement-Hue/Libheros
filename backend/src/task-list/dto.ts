
export class CreateTaskListDto {
  id: string;
  name: string;
}

export class CreateTaskDto {
  id: string;
  name: string;
  description: string;
  date: Date;
  creationDate: Date;
  completed?: boolean;
}