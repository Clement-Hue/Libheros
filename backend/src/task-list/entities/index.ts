export { default as Task } from './task.entity';
export { default as TaskList } from './taskList.entity';
