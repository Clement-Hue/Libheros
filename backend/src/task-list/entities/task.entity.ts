import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToOne,
  PrimaryColumn,
} from 'typeorm';
import TaskList from './taskList.entity';

@Entity()
export default class Task {
  @PrimaryColumn()
  id: string;

  @Column({ type: 'date' })
  date: Date;

  @Column({ type: 'datetime' })
  creationDate: Date;

  @Column()
  name: string;

  @Column({ type: 'text' })
  description: string;

  @Column({ default: false })
  completed: boolean;

  @ManyToOne(() => TaskList, (taskList) => taskList.tasks, {
    onDelete: 'CASCADE',
  })
  taskList: TaskList;
}
