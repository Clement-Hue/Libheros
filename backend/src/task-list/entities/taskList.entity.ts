import { Entity, PrimaryColumn, Column, OneToMany } from 'typeorm';
import Task from './task.entity';

@Entity()
export default class TaskList {
  @PrimaryColumn()
  id: string;

  @Column()
  name: string;

  @OneToMany(() => Task, (task) => task.taskList, { cascade: true })
  tasks: Task[];
}
