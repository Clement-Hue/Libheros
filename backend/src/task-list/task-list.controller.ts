import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  Request,
  Post, UseGuards,
} from '@nestjs/common';
import { TaskListService } from './task-list.service';
import { CreateTaskDto, CreateTaskListDto } from './dto';
import { AuthGuard } from '../auth/auth.guard';

@UseGuards(AuthGuard)
@Controller('task-list')
export class TaskListController {
  constructor(private readonly taskListService: TaskListService) {}

  @Post()
  @HttpCode(HttpStatus.CREATED)
  async createTaskList(@Body() dto: CreateTaskListDto, @Request() req) {
    const taskList = await this.taskListService.createTaskList({
      name: dto.name,
      id: dto.id,
      tasks: [],
      user: req.user
    });
    const { user, ...data } = taskList;
    return {
      data
    };
  }

  @Get()
  @HttpCode(HttpStatus.OK)
  async getTaskLists(@Request() req) {
    const taskList = await this.taskListService.getAllTaskList(req.user);
    return {
      data: taskList,
    };
  }

  @Get(':id')
  @HttpCode(HttpStatus.OK)
  async getTasks(@Param('id') listId: string, @Request() req) {
    const tasks = await this.taskListService.getTasksByListId(listId, req.user);
    return {
      data: tasks,
    };
  }

  @Post(':id')
  @HttpCode(HttpStatus.CREATED)
  async createTask(@Param('id') listId: string, @Body() dto: CreateTaskDto, @Request() req) {
    const task = await this.taskListService.addTaskToList(
      listId,
      req.user,
      dto,
    );
    return {
      data: task,
    };
  }

  @Delete(':id')
  @HttpCode(HttpStatus.OK)
  async deleteTaskList(@Param('id') listId: string, @Request() req) {
    await this.taskListService.removeTaskList(listId, req.user);
  }
}
