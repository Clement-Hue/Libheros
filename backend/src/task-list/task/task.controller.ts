import {
  Body,
  Controller,
  Delete,
  HttpCode,
  HttpStatus,
  Param,
  Put,
} from '@nestjs/common';
import { TaskListService } from '../task-list.service';
import { UpdateTaskDto } from '../dto';

@Controller('task')
export class TaskController {
  constructor(private readonly taskListService: TaskListService) {}

  @Delete(':id')
  @HttpCode(HttpStatus.OK)
  async removeTask(@Param('id') taskId: string) {
    await this.taskListService.removeTask(taskId);
  }

  @Put(':id')
  @HttpCode(HttpStatus.OK)
  async updateTask(@Param('id') taskId: string, @Body() data: UpdateTaskDto) {
    const task = await this.taskListService.updateTask(taskId, data);
    return {
      data: task,
    };
  }
}
