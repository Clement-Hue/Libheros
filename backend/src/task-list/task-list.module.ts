import { Module } from '@nestjs/common';
import { TaskListController } from './task-list.controller';
import { TaskListService } from './task-list.service';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Task, TaskList } from './entities';
import { TaskController } from './task/task.controller';

@Module({
  imports: [TypeOrmModule.forFeature([Task, TaskList])],
  controllers: [TaskListController, TaskController],
  providers: [TaskListService],
})
export class TaskListModule {}
