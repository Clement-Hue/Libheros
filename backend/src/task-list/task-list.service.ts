import { Injectable, NotFoundException } from '@nestjs/common';
import { Repository } from 'typeorm';
import { InjectRepository } from '@nestjs/typeorm';
import { Task, TaskList } from './entities';

@Injectable()
export class TaskListService {
  constructor(
    @InjectRepository(TaskList)
    private taskListRepository: Repository<TaskList>,
    @InjectRepository(Task)
    private taskRepository: Repository<Task>,
  ) {}

  getAllTaskList(): Promise<TaskList[]> {
    return this.taskListRepository.find({
      relations: ['tasks'],
    });
  }

  async removeTaskList(id: string): Promise<void> {
    const taskList = await this.taskListRepository.findOne({
      where: { id },
    });
    if (!taskList) {
      throw new NotFoundException({ code: 'list.no-found' });
    }
    await this.taskListRepository.delete(id);
  }

  async getTasksByListId(taskListId: string) {
    return this.taskRepository.find({
      where: { taskList: { id: taskListId } },
    });
  }

  async createTaskList(data: TaskList): Promise<TaskList> {
    const taskList = this.taskListRepository.create(data);
    return this.taskListRepository.save(taskList);
  }

  async addTaskToList(
    listId: string,
    taskData: Omit<Task, 'taskList'>,
  ): Promise<Task> {
    const list = await this.taskListRepository.findOne({
      where: { id: listId },
    });
    if (!list) throw new NotFoundException({ code: 'list.no-found' });

    const task = this.taskRepository.create({ ...taskData, taskList: list });
    return this.taskRepository.save(task);
  }

  async updateTask(
    taskId: string,
    updates: Omit<Partial<Task>, 'taskList'>,
  ): Promise<Task> {
    const task = await this.taskRepository.findOne({
      where: { id: taskId },
    });

    if (!task) throw new NotFoundException({ code: 'task.no-found' });

    Object.assign(task, updates);
    return this.taskRepository.save(task);
  }

  async removeTask(taskId: string): Promise<void> {
    const task = await this.taskRepository.findOne({
      where: { id: taskId },
    });

    if (!task) throw new NotFoundException({ code: 'task.not-found' });

    await this.taskRepository.remove(task);
  }
}
