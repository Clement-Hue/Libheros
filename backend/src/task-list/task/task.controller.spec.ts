import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import { TaskListService } from '../task-list.service';
import { Repository } from 'typeorm';
import { Task, TaskList } from '../entities';
import {
  createTestingApp, createUserAndToken,
  TestingModules,
} from '../../test-utils/testing-modules';
import { getRepositoryToken } from '@nestjs/typeorm';
import * as request from 'supertest';
import { TaskController } from './task.controller';
import { AuthModule } from '../../auth/auth.module';

const email = 'pierre@hotmail.fr';

describe('TaskController', () => {
  let app: INestApplication;
  let taskRepository: Repository<Task>;
  let taskListRepository: Repository<TaskList>;
  let token: string

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [...TestingModules(), AuthModule],
      controllers: [TaskController],
      providers: [TaskListService],
    }).compile();

    app = await createTestingApp(module);
    taskListRepository = module.get<Repository<TaskList>>(
      getRepositoryToken(TaskList),
    );
    taskRepository = module.get<Repository<Task>>(getRepositoryToken(Task));
    token = await createUserAndToken(module)({ email }).then(
      ({ token }) => token,
    );
  });

  afterAll(async () => {
    await app.close();
  });

  it('should remove task', async () => {
    const task = {
      id: 'task-1',
      name: 'Task A',
      description: '',
      creationDate: new Date('2025-02-01').toISOString(),
      date: new Date('2025-02-10').toISOString(),
      completed: false,
    };
    const list = {
      id: 'id-1',
      name: 'my list 1',
      tasks: [task],
      user: { email },
    };
    await taskListRepository.save(list);
    expect(
      await taskRepository.findOne({
        where: {
          id: 'task-1',
        },
      }),
    ).not.toBeNull();
    await request(app.getHttpServer()).delete('/task/task-1')
      .set('Authorization', `Bearer ${token}`)
      .expect(200);
    expect(
      await taskRepository.findOne({
        where: {
          id: 'task-1',
        },
      }),
    ).toBeNull();
  });

  it('should show error while updating task', async () => {
    const res = await request(app.getHttpServer())
      .put('/task/task-1')
      .set('Authorization', `Bearer ${token}`)
      .send({
        name: 'Task AB',
        description: '',
        creationDate: new Date('2025-02-01').toISOString(),
        date: new Date('2025-02-10').toISOString(),
        completed: true,
      })
      .expect(404);
    expect(res.body).toEqual({
      code: 'task.not-found',
    });
  });

  it('should update task', async () => {
    const task = {
      id: 'task-1',
      name: 'Task A',
      description: '',
      creationDate: new Date('2025-02-01').toISOString(),
      date: new Date('2025-02-10').toISOString(),
      completed: false,
    };
    const list = {
      id: 'id-1',
      name: 'my list 1',
      tasks: [task],
      user: { email },
    };
    await taskListRepository.save(list);
    const res = await request(app.getHttpServer())
      .put('/task/task-1')
      .set('Authorization', `Bearer ${token}`)
      .send({
        name: 'Task AB',
        description: '',
        creationDate: new Date('2025-02-01').toISOString(),
        date: new Date('2025-02-10').toISOString(),
        completed: true,
      })
      .expect(200);
    expect(res.body).toMatchSnapshot();
  });

  it('should show unauthorized', async () => {
    await request(app.getHttpServer())
      .put('/task/task-1')
      .send({
        name: 'Task AB',
        description: '',
        creationDate: new Date('2025-02-01').toISOString(),
        date: new Date('2025-02-10').toISOString(),
        completed: true,
      })
      .expect(401);
  });
});
