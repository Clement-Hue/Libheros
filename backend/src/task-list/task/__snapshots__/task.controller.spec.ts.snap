// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskController should update task 1`] = `
{
  "data": {
    "completed": true,
    "creationDate": "2025-02-01T00:00:00.000Z",
    "date": "2025-02-10T00:00:00.000Z",
    "description": "",
    "id": "task-1",
    "name": "Task AB",
  },
}
`;
